.circleci/config.yml

```diff
@@ -680,6 +624,56 @@ jobs:
           command: |
             gh-pages --dotfiles --message "[skip ci] documentation update" --dist target/doc
 
+  ######################################################################################################
+  # Test building docker ci base image if relevnat files have changed                                  #
+  ######################################################################################################
+  docker-ci-build:
+    executor: docker-executor
+    description: publish docker images
+    steps:
+      - checkout
+      - setup_docker_signing
+      - run:
+          name: Build and publish the ci base image.
+          command: |
+
+            #detect if any relevant files have changed requiring base docker file rebuilding.
+            #this is specific for prs targeting master.   TODO: figure out target release branch.
+            git --no-pager diff --name-only HEAD $(git merge-base HEAD origin/master) | sort > /tmp/changed_files
+            echo ".circleci/config.yml" >> /tmp/ci_docker_files
+            echo "cargo-toolchain" >> /tmp/ci_docker_files
+            echo "docker/ci/Dockerfile" >> /tmp/ci_docker_files
+            echo "rust-toolchain" >> /tmp/ci_docker_files
+            echo "scripts/dev_setup.sh" >> /tmp/ci_docker_files
+
+            #if no relevant files have changed - exit
+            if [ `join /tmp/changed_files /tmp/ci_docker_files | wc -l` == 0 ]; then
+              echo no relevant files have changed - exit
+              exit 0
```

Comment:
Consider the alternative `circleci step halt` in this case, for example, if we plan to have more steps in this job in future.

Reply:
done